### PR TITLE
Fix(room): Calculate display_name correctly for invited rooms

### LINKED ISSUE #133: Empty rooms get an empty string as the display name

### PR DESCRIPTION
**Pull request by gnunicorn**
- [x] Make the `DisplayName` an actual type to allow API users to localize the generic name
 - [x] tests for various naming use cases
 - [x] fix the cases where we currently calculate a faulty name, fixes #133 
 - [x] strong-typed EitherMemberEvent from storage
 - [x] add tests for eithermember in store integration tests 
 - [x] update sled store for new trait signature
 - [x] update indexeddb store for new signature
 - [x] rename `DisplayName::Computed` to `DisplayName::Calculated`
 - [x] fix renaming failing test(s)

### CODE REVIEW

File: crates/matrix-sdk-common/src/deserialized_responses.rs [hunk 1]
```diff
@@ -290,6 +296,43 @@ impl TimelineSlice {
     }
 }
 
+#[allow(clippy::large_enum_variant)]
+/// Wrapper around both MemberEvent-Types
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub enum MemberEvent {
+    Stripped(StrippedRoomMemberEvent),
+    Full(OriginalSyncRoomMemberEvent),
```
Comment (jplatte) [inline crates/matrix-sdk-common/src/deserialized_responses.rs:304]: Why use a different naming than Ruma here? "Full" is used in a few other places to mean "including the room_id unlike events from sync".

Also, what about redacted ones? Could deal with that in a separate PR, but it's going to significantly alter what you added.

File: crates/matrix-sdk-common/src/deserialized_responses.rs [hunk 2]
```diff
@@ -290,6 +296,43 @@ impl TimelineSlice {
     }
 }
 
+#[allow(clippy::large_enum_variant)]
+/// Wrapper around both MemberEvent-Types
```
Comment (jplatte) [inline crates/matrix-sdk-common/src/deserialized_responses.rs:300]:
```suggestion
/// Wrapper around both MemberEvent-Types
#[allow(clippy::large_enum_variant)]
```

File: crates/matrix-sdk-common/src/deserialized_responses.rs (hunk 1 above)
Comment (gnunicorn) [reply crates/matrix-sdk-common/src/deserialized_responses.rs:304]: > Why use a different naming than Ruma here? "Full" is used in a few other places to mean "including the room_id unlike events from sync".

You mean going for `Original` instead of `Full`? I don't really care. And I don't think it really matters (as this is mostly internal anyways).

> Also, what about redacted ones? Could deal with that in a separate PR, but it's going to significantly alter what you added.

Because this tries to fix the name calculation only - the PR is already a lot more complex than I expected this fix to be needed for. I don't really want to add more complexity by adding yet another aspect.

Sure, the other PR will alter this, that's fine. But after this PR we have tests in place that ensure whatever is being done, this problem stays fixed.

File: crates/matrix-sdk-sled/src/state_store.rs [hunk 3]
```diff
@@ -507,6 +507,28 @@ impl SledStore {
 
                     for (room, events) in &changes.stripped_members {
                         for event in events.values() {
+                            let key = (room, &event.state_key);
+
+                            match event.content.membership {
+                                MembershipState::Join => {
+                                    joined.insert(
```
Comment (poljar) [inline crates/matrix-sdk-sled/src/state_store.rs:514]: I think this is problematic. This is now mixing up user ids that we pick out from the normal state with those we pick out from the stripped state.

This means that we could start making E2EE related decisions based on information coming from the stripped state. The list of joined and invited user ids is directly used to decide which members get the current room key.

Stripped state is by definition less trusted:

>  Although stripped state is usually generated and provided by the server, it is still possible to be incorrect on the receiving end. The stripped state events are not signed and could theoretically be modified, or outdated due to updates not being sent.

There's also this in the spec:
> Clients should only use stripped state events so long as they don’t have access to the proper state of the room. Once the state of the room is available, all stripped state should be discarded. In cases where the client has an archived state of the room (such as after being kicked) and the client is receiving stripped state for the room, such as from an invite or knock, then the stripped state should take precedence until fresh state can be acquired from a join.

While I don't think we need to discard it completely, we certainly can't mix things from stripped state and normal state.

Link to the relevant spec entry: https://spec.matrix.org/v1.2/client-server-api/#stripped-state

**Review by poljar — CHANGES_REQUESTED** [review]
Thanks for the comprehensive test improvements around this. Most of this looks good and sane, sadly I don't think we can mix data from stripped state in this way.

The top level method to fetch users can stay the same, but the underlying data needs to be kept in separate buckets.

File: crates/matrix-sdk-sled/src/state_store.rs (hunk 3 above)
Comment (gnunicorn) [reply crates/matrix-sdk-sled/src/state_store.rs:514]: Thanks for sharing that insight, I wasn't aware about this aspect.

> The list of joined and invited user ids is directly used to decide which members get the current room key.

But `UserId` isn't actually enough to share that information with, right? They are only an early indicator about the user_ids present in a room. So in the real-world scenario, we'd still be fetching the actual `Member`-Object, which indicates whether that information was given from stripped or not (or at least can tell). So, in reality this would still be okay-ish.

However
>  Once the state of the room is available, all stripped state should be discarded.

If that is an assumption made by spec, then even data extracted from the stripped state is probably necessary to be deleted. While we do delete the user_id from the other index, whenever we see a new membership event after, we don't necessarily remove them properly when we were told one in stripped_state that never actually happened. While practically unlikely to cause any problem, it would still leave us in an inconsistent state.

Okay. So, what is the proposed solution here then: second set of indizes for the stripped_states (that we purge any room info when we see it on the regular room) and do a fallback lookup on them the same way we handle the member info itself right now?

Comment (poljar) [reply crates/matrix-sdk-sled/src/state_store.rs:514]: > But `UserId` isn't actually enough to share that information with, right? They are only an early indicator about the user_ids present in a room. So in the real-world scenario, we'd still be fetching the actual `Member`-Object, which indicates whether that information was given from stripped or not (or at least can tell). So, in reality this would still be okay-ish.

Actually the `UserId` is enough, well as far as the member event is concerned. The joined/invited/left user ids lists are a store optimization. 

The state-store trait has a method to get the joined and invited user ids. How this is implemented is left to decide for each implementation. If they want to deserialize all member events and count that or if they have an index as a optimization doesn't matter. It does need to be correct and match to what the member events have.

If they do not match you'll risk having user ids for which you do not have the matching event which would lead to the client rendering one list of users while another list of users gets room keys.

> If that is an assumption made by spec, then even data extracted from the stripped state is probably necessary to be deleted. While we do delete the user_id from the other index, whenever we see a new membership event after, we don't necessarily remove them properly when we were told one in stripped_state that never actually happened. While practically unlikely to cause any problem, it would still leave us in an inconsistent state.

Precisely, this inconsistent state would lead to the above mentioned security vulnerability.

> Okay. So, what is the proposed solution here then: second set of indices for the stripped_states (that we purge any room info when we see it on the regular room) and do a fallback lookup on them the same way we handle the member info itself right now?

Correct, never mix events or info from stripped state with normal state. Rooms that are considered joined/left pick data from the normal buckets. Rooms that are considered invited/knocked pick data from the stripped state buckets.

File: crates/matrix-sdk-sled/src/state_store.rs [hunk 4]
```diff
@@ -426,6 +429,12 @@ impl SledStore {
                                 self.serialize_event(&event)
                                     .map_err(ConflictableTransactionError::Abort)?,
                             )?;
+                            // clear out any remaining stripped data
+                            stripped_members.remove(self.encode_key(STRIPPED_ROOM_MEMBER, &key))?;
```
Comment (poljar) [inline crates/matrix-sdk-sled/src/state_store.rs:433]: Are we starting to discard stripped state because of this sentence:

> Clients should only use stripped state events so long as they don’t have access to the proper state of the room. Once the state of the room is available, all stripped state should be discarded. In cases where the client has an archived state of the room (such as after being kicked) and the client is receiving stripped state for the room, such as from an invite or knock, then the stripped state should take precedence until fresh state can be acquired from a join.

If so, what about `striped_rooms` and `stripped_state`.

File: crates/matrix-sdk-sled/src/state_store.rs [hunk 5]
```diff
@@ -505,14 +498,41 @@ impl SledStore {
...
-                                self.encode_key(
-                                    STRIPPED_ROOM_MEMBER,
-                                    (room, event.state_key.to_string()),
-                                ),
+                                self.encode_key(STRIPPED_ROOM_MEMBER, &key),
                                 self.serialize_event(&event)
                                     .map_err(ConflictableTransactionError::Abort)?,
                             )?;
+
+                            members.remove(self.encode_key(MEMBER, &key))?;
```
Comment (poljar) [inline crates/matrix-sdk-sled/src/state_store.rs:533]: This also just touches the members while it ignores other things.

Also I'm not sure about this. Our room becomes archived and we serve the state from the stripped buckets, but the spec mentions, or used to mention, this:

> If the client joins the room then the current state will be given as a delta against the archived state not the invite_state.

Now this is related to this issue: https://github.com/matrix-org/matrix-spec/issues/814

Synapse does return the whole state once you rejoin a left room so I don't think that this will break anything, though I would still just leave it be. It does protect us from e2ee downgrade attacks between rejoins.

File: crates/matrix-sdk-sled/src/state_store.rs [hunk 6]
```diff
@@ -585,6 +613,26 @@ impl SledStore {
 
         self.save_room_timeline(changes).await?;
 
+        // user state
+        let ret: Result<(), TransactionError<SledStoreError>> = (&self.session, &self.account_data)
```
Comment (poljar) [inline crates/matrix-sdk-sled/src/state_store.rs:617]: We now have 3 transactions? Doesn't this fit into the previous one like presence did?

**Review by poljar — CHANGES_REQUESTED** [review]

File: crates/matrix-sdk-sled/src/state_store.rs (hunk 6 above)
Comment (gnunicorn) [reply crates/matrix-sdk-sled/src/state_store.rs:617]: I tried splitting them around semantics after not being able to put more entries into the existing one. Might be able to coble that into two, but not quite the right split. But I guess either is bad as it leaves incomplete state, if any one fails... so my choice was to at least have them semantically grouped...

File: crates/matrix-sdk-sled/src/state_store.rs (hunk 4 above)
Comment (gnunicorn) [reply crates/matrix-sdk-sled/src/state_store.rs:433]: This was removed as we have, on a higher level, a mixing of looking up that data. To ensure we don't accidentally return bad stale state, I added some code here to clear the opposing index.

File: crates/matrix-sdk-sled/src/state_store.rs (hunk 5 above)
Comment (gnunicorn) [reply crates/matrix-sdk-sled/src/state_store.rs:533]: Hmm... this honestly becomes a lot of logic for a rather simple kv-store to manage, and very easy for an implementation to not adhere to proper. We might have a bunch of integration tests, but many of these aspects are not tested for, and the assumption that an implementer knows about this edge cases of the matrix spec sounds daunting to me. I think we should rethink the API and where the logic should lie for state stores going forward...

Comment (poljar) [reply crates/matrix-sdk-sled/src/state_store.rs:533]: I think we can just skipping the removal here, not removing the state her is fine:
1. If servers return the full state, it will just overwrite the old ones
2. If servers return the delta from the last time we were joined

So I don't think we need any additional logic here.

File: crates/matrix-sdk-sled/src/state_store.rs (hunk 6 above)
Comment (poljar) [reply crates/matrix-sdk-sled/src/state_store.rs:617]: I guess as long as the sync token is persisted last we're fineish, no matter how many transactions we have.

File: crates/matrix-sdk-sled/src/state_store.rs (hunk 5 above)
Comment (gnunicorn) [reply crates/matrix-sdk-sled/src/state_store.rs:533]: you remove the current removing code? or leave it as is?

Comment (poljar) [reply crates/matrix-sdk-sled/src/state_store.rs:533]: I mean remove the current removing code.

File: crates/matrix-sdk-base/src/store/memory_store.rs [hunk 7]
```diff
@@ -243,6 +243,39 @@ impl MemoryStore {
 
         for (room, events) in &changes.stripped_members {
             for event in events.values() {
+                match event.content.membership {
+                    MembershipState::Join => {
+                        self.joined_user_ids
```
Comment (poljar) [inline crates/matrix-sdk-base/src/store/memory_store.rs:248]: Isn't this still mixing stripped user ids and non-stripped ones?

File: crates/matrix-sdk-indexeddb/src/state_store.rs [hunk 8]
```diff
@@ -473,10 +476,36 @@ impl IndexeddbStore {
...
+            let invited = tx.object_store(KEYS::INVITED_USER_IDS)?;
             for (room, events) in &changes.stripped_members {
                 for event in events.values() {
-                    let key = self.encode_key(KEYS::STRIPPED_MEMBERS, (room, &event.state_key));
-                    store.put_key_val(&key, &self.serialize_event(&event)?)?;
+                    let key = (room, &event.state_key);
+
+                    match event.content.membership {
+                        MembershipState::Join => {
+                            joined.put_key_val_owned(
```
Comment (poljar) [inline crates/matrix-sdk-indexeddb/src/state_store.rs:487]: Same question, this is still mixing the states?

**Review by poljar — CHANGES_REQUESTED** [review]
The non-sled stores seem to be still mixing things or am I missing something?

File: crates/matrix-sdk-base/src/store/memory_store.rs (hunk 7 above)
Comment (gnunicorn) [reply crates/matrix-sdk-base/src/store/memory_store.rs:248]: sorry, you are right.

**Review by poljar — APPROVED** [review]
